<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Identify spatially variable genes by projecting to morphologically relevant curves. • MorphoGAM</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Identify spatially variable genes by projecting to morphologically relevant curves.">
<meta name="description" content='Includes methods (interactive and automatic) for estimating a "morphologically relevant" curve from spatial transcriptomics data. Implements a generalized additive model (GAM) to identify gene expression variation along the curve.'>
<meta property="og:description" content='Includes methods (interactive and automatic) for estimating a "morphologically relevant" curve from spatial transcriptomics data. Implements a generalized additive model (GAM) to identify gene expression variation along the curve.'>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">MorphoGAM</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/getting-started.html">getting-started</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><p>MorphoGAM: Detect spatially variable genes by projecting to morphologically relevant curves ================ R package version 1.1.0</p>
<div class="section level2">
<h2 id="system-requirements">System Requirements<a class="anchor" aria-label="anchor" href="#system-requirements"></a>
</h2>
<p><code>R</code> is required to use <code>MorphoGAM</code>. In development, <code>R</code> version 4.3.0 and greater were used, but there may be compatibility with previous versions.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>From the R console, <code>devtools::install_github("phillipnicol/MorphoGAM")</code>. Installation should take less than a minute on a standard machine.</p>
</div>
<div class="section level2">
<h2 id="demo">Demo<a class="anchor" aria-label="anchor" href="#demo"></a>
</h2>
<p>This demonstration should complete in a few minutes on a standard computer.</p>
<div class="section level3">
<h3 id="step-1-estimate-morphologically-relevant-coordinates">Step 1: Estimate morphologically relevant coordinates<a class="anchor" aria-label="anchor" href="#step-1-estimate-morphologically-relevant-coordinates"></a>
</h3>
<p>The first step in running <code>MorphoGAM</code> is to define a curve from which the morphologically relevant coordinates are defined. To demonstrate this, we use the swiss roll example:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://phillipnicol.github.io/MorphoGAM/">MorphoGAM</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="va">xy</span> <span class="op">&lt;-</span> <span class="fu">MorphoGAM</span><span class="fu">:::</span><span class="fu">makeSwissRoll</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">xy</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>,y<span class="op">=</span><span class="va">xy</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>,y<span class="op">=</span><span class="va">y</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>size<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/unnamed-chunk-1-1.png"><!-- --></p>
<p>The function <code><a href="reference/CurveFinder.html">CurveFinder()</a></code> applies the automatic curve estimation method</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/CurveFinder.html">CurveFinder</a></span><span class="op">(</span><span class="va">xy</span><span class="op">)</span></span></code></pre></div>
<p>The <code>fit</code> object contains plots of the first two morphologically relevant coordinates and the fitted curve:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span><span class="op">$</span><span class="va">curve.plot</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/unnamed-chunk-3-1.png"><!-- --></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span><span class="op">$</span><span class="va">coordinate.plot</span> <span class="co">#First morphologically relevant coordinate </span></span></code></pre></div>
<p><img src="README_files/figure-gfm/unnamed-chunk-3-2.png"><!-- --></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span><span class="op">$</span><span class="va">residuals.plot</span> <span class="co">#Second morphologically relevant coordinate</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/unnamed-chunk-3-3.png"><!-- --></p>
<p>The morphologically relevant coordintes can be accessed via <code>fit$xyt</code>:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span><span class="op">$</span><span class="va">xyt</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="do">##            x          y         t         r         f1         f2</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="do">## 1 -0.5570179  0.3213394 0.2050271 0.4133826 -0.5474196  0.3125001</span></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="do">## 2 -0.5829042 -0.4035348 0.2980582 0.4115648 -0.5730570 -0.3944718</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="do">## 3  0.6330314 -0.3873620 0.4895029 0.7098784  0.6685151 -0.4089267</span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="do">## 4 -0.8982847  0.2923785 0.8824209 0.6039158 -0.9116261  0.3098966</span></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a><span class="do">## 5 -0.2294434  0.5582292 0.1530027 0.4402809 -0.2343397  0.5517795</span></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a><span class="do">## 6 -0.8106100  0.3995968 0.8657898 0.6428067 -0.8339070  0.4171639</span></span></code></pre></div>
<p>In some cases the user may wish to draw the curve by hand. For this we provide an interactive shiny app that can be run locally using the function <code><a href="reference/CurveFinderInteractive.html">CurveFinderInteractive()</a></code>. Once the app is running you can click the sequence of points defining the curve, then press the “Smooth” button to fit the curve. Once the smoothing is done the app can be closed and <code>fit</code> will be returned.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="co">#Running this opens a shiny app</span></span>
<span>  <span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/CurveFinderInteractive.html">CurveFinderInteractive</a></span><span class="op">(</span><span class="va">xy</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-2-apply-gam-to-morphologically-relevant-coordinates">Step 2: Apply GAM to morphologically relevant coordinates<a class="anchor" aria-label="anchor" href="#step-2-apply-gam-to-morphologically-relevant-coordinates"></a>
</h3>
<p>The next step is to identify genes with variable expression along the curve (or in the orthogonal direction). Here we generate a synthetic count matrix <code>Y</code> with one spatially interesting gene:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">100</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">xy</span><span class="op">)</span>, lambda<span class="op">=</span><span class="fl">1</span><span class="op">)</span>,</span>
<span>            nrow<span class="op">=</span><span class="fl">100</span>, ncol<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">xy</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">eta</span> <span class="op">&lt;-</span> <span class="op">-</span><span class="fl">3</span><span class="op">*</span><span class="va">fit</span><span class="op">$</span><span class="va">xyt</span><span class="op">$</span><span class="va">t</span> <span class="op">+</span> <span class="fl">2</span></span>
<span><span class="va">Y</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">xy</span><span class="op">)</span>,lambda<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="va">eta</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Gene"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Now we apply the generalized additive model (GAM):</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mgam</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/MorphoGAM.html">MorphoGAM</a></span><span class="op">(</span><span class="va">Y</span>, curve.fit<span class="op">=</span><span class="va">fit</span>,</span>
<span>                  design <span class="op">=</span> <span class="va">y</span> <span class="op">~</span> <span class="fu">s</span><span class="op">(</span><span class="va">t</span>, bs<span class="op">=</span><span class="st">"cr"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="do">## ================================================================================</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="do">## Warning in irlba::irlba(fxs.r): fast code path error starting vector near the</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a><span class="do">## null space; re-trying with fastpath=FALSE.</span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a><span class="do">## Error in irlba::irlba(fxs.r) : starting vector near the null space</span></span></code></pre></div>
<p>The <code>bs = "cr"</code> specifies cubic regression splines in the GAM, although this can be modified to periodic splines or other basis functions provided by <code>mgcv</code>. We may wish to sort the results matrix to rank genes by summaries of the estimated function:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mgam</span><span class="op">$</span><span class="va">results</span> <span class="op">|&gt;</span> <span class="fu">arrange</span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">peak.t</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="do">##             peak.t    range.t         pv.t peak.r range.r pv.r intercept</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="do">## Gene 1  1.18349653 5.40344926 0.0000000000      0       0    0 -3.953473</span></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a><span class="do">## Gene 33 0.10195765 0.18728675 0.0002472835      0       0    0 -4.618065</span></span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a><span class="do">## Gene 20 0.06313620 0.11151799 0.0063965136      0       0    0 -4.647023</span></span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a><span class="do">## Gene 49 0.04881890 0.05290388 0.0506582449      0       0    0 -4.682800</span></span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a><span class="do">## Gene 16 0.03538251 0.04356503 0.0392955411      0       0    0 -4.610617</span></span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a><span class="do">## Gene 10 0.03233560 0.03115972 0.2185367051      0       0    0 -4.645891</span></span></code></pre></div>
<p>The results indicate gene <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>1</mn><annotation encoding="application/x-tex">1</annotation></semantics></math> has a significant peak and range (region of increased expression), and we can visually confirm this by using <code>plotGAMestimate()</code> to plot the entire function:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plotGAMestimates.html">plotGAMestimates</a></span><span class="op">(</span><span class="va">Y</span>,genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene 1"</span>, <span class="st">"Gene 2"</span><span class="op">)</span>,</span>
<span>                 mgam_object <span class="op">=</span> <span class="va">mgam</span>,</span>
<span>                 curve_fit<span class="op">=</span><span class="va">fit</span>,</span>
<span>                 type<span class="op">=</span><span class="st">"t"</span><span class="op">)</span></span></code></pre></div>
<p><img src="README_files/figure-gfm/unnamed-chunk-9-1.png"><!-- --></p>
<p>To also identify genes that vary in the direction of the second morphologically relevant coordinate, add the term <code>s(r, ...)</code> to the <code>design</code> argument in <code>MorphoGAM</code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mgam_with_r</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/MorphoGAM.html">MorphoGAM</a></span><span class="op">(</span><span class="va">Y</span>, curve.fit<span class="op">=</span><span class="va">fit</span>,</span>
<span>                          design <span class="op">=</span> <span class="va">y</span> <span class="op">~</span> <span class="fu">s</span><span class="op">(</span><span class="va">t</span>, bs<span class="op">=</span><span class="st">"cr"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">s</span><span class="op">(</span><span class="va">r</span>, bs<span class="op">=</span><span class="st">"cr"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="reference">Reference<a class="anchor" aria-label="anchor" href="#reference"></a>
</h2>
<p>If you use <code>MorphoGAM</code> in your work, please cite:</p>
</div>

  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing MorphoGAM</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Phillip B. Nicol <br><small class="roles"> Maintainer </small>   </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Phillip B. Nicol.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
